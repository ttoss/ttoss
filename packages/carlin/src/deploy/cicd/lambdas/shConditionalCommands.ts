import type { Status } from './ecsTaskReport.handler';

export const compileCommands = (commands: string[]) => {
  return commands.map((c) => c.replace(/;$/, '')).join(' && ');
};

const approvedStatus: Status = 'Approved';

const rejectedStatus: Status = 'Rejected';

const defaultSuccessCommands = [
  `carlin cicd-ecs-task-report --status=${approvedStatus}`,
];

const defaultFailureCommands = [
  `carlin cicd-ecs-task-report --status=${rejectedStatus}`,
];

const defaultFinallyCommands = ['echo "Finally Command"'];

export const shConditionalCommands = ({
  conditionalCommands,
  successCommands = defaultSuccessCommands,
  failureCommands = defaultFailureCommands,
  finallyCommands = defaultFinallyCommands,
}: {
  conditionalCommands: string[];
  successCommands?: string[];
  failureCommands?: string[];
  finallyCommands?: string[];
}) => {
  const conditionalCommand = compileCommands(conditionalCommands);

  const successCommand = compileCommands([
    'echo "Success Command"',
    ...successCommands,
  ]);

  const failureCommand = compileCommands([
    'echo "Failure Command"',
    ...failureCommands,
  ]);

  const finallyCommand = compileCommands(finallyCommands);

  return `if ${conditionalCommand}; then ${successCommand}; else ${failureCommand}; fi; ${finallyCommand}`;
};
