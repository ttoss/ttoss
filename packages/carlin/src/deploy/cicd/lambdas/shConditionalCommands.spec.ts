import { shConditionalCommands } from './shConditionalCommands';

test.each([
  {
    conditionalCommands: ['conditional command 1', 'conditional command 2'],
  },
  {
    conditionalCommands: ['conditional command 1', 'conditional command 2'],
    successCommands: ['success command 1', 'success command 2'],
  },
  {
    conditionalCommands: ['conditional command 1', 'conditional command 2'],
    successCommands: ['success command 1', 'success command 2'],
    failureCommands: ['failure command 1', 'failure command 2'],
  },
  {
    conditionalCommands: ['conditional command 1', 'conditional command 2'],
    successCommands: ['success command 1', 'success command 2'],
    failureCommands: ['failure command 1', 'failure command 2'],
    finallyCommands: ['finally command 1', 'finally command 2'],
  },
])('should compile commands %#', (commands) => {
  const compiledCommands = shConditionalCommands(commands);

  expect(compiledCommands).toMatchSnapshot();
});
