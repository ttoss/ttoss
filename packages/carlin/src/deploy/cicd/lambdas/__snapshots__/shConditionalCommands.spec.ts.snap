// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should compile commands 0 1`] = `"if conditional command 1 && conditional command 2; then echo "Success Command" && carlin cicd-ecs-task-report --status=Approved; else echo "Failure Command" && carlin cicd-ecs-task-report --status=Rejected; fi; echo "Finally Command""`;

exports[`should compile commands 1 1`] = `"if conditional command 1 && conditional command 2; then echo "Success Command" && success command 1 && success command 2; else echo "Failure Command" && carlin cicd-ecs-task-report --status=Rejected; fi; echo "Finally Command""`;

exports[`should compile commands 2 1`] = `"if conditional command 1 && conditional command 2; then echo "Success Command" && success command 1 && success command 2; else echo "Failure Command" && failure command 1 && failure command 2; fi; echo "Finally Command""`;

exports[`should compile commands 3 1`] = `"if conditional command 1 && conditional command 2; then echo "Success Command" && success command 1 && success command 2; else echo "Failure Command" && failure command 1 && failure command 2; fi; finally command 1 && finally command 2"`;
